File: app/matestack/bootstrap/components/chart.js
import Chart from 'chart.js';



MatestackUiCore.Vue.component('matestack-ui-bootstrap-chart', {
  mixins: [MatestackUiCore.componentMixin],

  data() {
    return {
      chartJsInstance: undefined,
      defaultColor: undefined,
      fontColor: undefined,
      fontFamily: undefined
    };
  },

  methods: {
    getThemeColor: function(key){
      const style = getComputedStyle(document.body);
      return style.getPropertyValue('--bs-'+key);
    },
    getThemeColorArray: function(keysArray){
      let result = []
      const style = getComputedStyle(document.body);
      keysArray.forEach(function(key){
        result.push(style.getPropertyValue('--bs-'+key))
      })
      return result;
    },
    drawBarChart: function(chartElement){
      const self = this;
      this.componentConfig["datasets"].forEach(function(item){
        if (item["backgroundColor"] === undefined){
          item["backgroundColor"] = self.getThemeColor("primary")
        }else{
          if(Array.isArray(item["backgroundColor"])){
            item["backgroundColor"] = self.getThemeColorArray(item["backgroundColor"])
          }else{
            item["backgroundColor"] = self.getThemeColor(item["backgroundColor"])
          }
        }
        item["hoverBackgroundColor"] = "rgba(0, 0, 0, 0.1)"
        if (item["barThickness"] === undefined){
          item["barThickness"] = 10;
        }
      })
      this.chartJsInstance = new Chart(chartElement, {
        type: 'bar',
        data: {
            labels: this.componentConfig["labels"],
            datasets: this.componentConfig["datasets"]
        },
        options: {
            legend: {
              display: this.componentConfig["display_legend"],
            },
            scales: {
                yAxes: [{
                    display: this.componentConfig["display_y_axes"],
                    gridLines: {
                      display: false,
                    },
                    ticks: {
                        beginAtZero: true
                    }
                }],
                xAxes: [{
                    display: this.componentConfig["display_x_axes"],
                    gridLines: {
                      display: false,
                    },
                    ticks: {
                        beginAtZero: true
                    }
                }]
            }
        }
      });
    },
    drawLineChart: function(chartElement){
      const self = this;
      this.componentConfig["datasets"].forEach(function(item){
        if (item["borderColor"] === undefined){
          item["borderColor"] = self.getThemeColor("primary")
        }else{
          item["borderColor"] = self.getThemeColor(item["borderColor"])
        }
        item["hoverBackgroundColor"] = "rgba(0, 0, 0, 0.1)"
        if (item["fill"] === undefined){
          item["fill"] = false;
        }
      })
      this.chartJsInstance = new Chart(chartElement, {
        type: 'line',
        data: {
            labels: this.componentConfig["labels"],
            datasets: this.componentConfig["datasets"]
        },
        options: {
            legend: {
              display: this.componentConfig["display_legend"],
            },
            scales: {
                yAxes: [{
                    display: this.componentConfig["display_y_axes"],
                    gridLines: {
                      display: false,
                    },
                    ticks: {
                        beginAtZero: true
                    }
                }],
                xAxes: [{
                    display: this.componentConfig["display_x_axes"],
                    gridLines: {
                      display: false,
                    },
                    ticks: {
                        beginAtZero: true
                    }
                }]
            }
        }
      });
    },
    drawDoughnutChart: function(chartElement){
      const self = this;
      this.componentConfig["datasets"].forEach(function(item){
        if (item["backgroundColor"] === undefined){
          item["backgroundColor"] = self.getThemeColor("primary")
        }else{
          if(Array.isArray(item["backgroundColor"])){
            item["backgroundColor"] = self.getThemeColorArray(item["backgroundColor"])
          }else{
            item["backgroundColor"] = self.getThemeColor(item["backgroundColor"])
          }
        }
        item["hoverBackgroundColor"] = "rgba(0, 0, 0, 0.1)"
      })
      this.chartJsInstance = new Chart(chartElement, {
        type: 'doughnut',
        data: {
            labels: this.componentConfig["labels"],
            datasets: this.componentConfig["datasets"]
        },
        options: {
            legend: {
              display: this.componentConfig["display_legend"],
            }
        }
      });
    },
    drawPieChart: function(chartElement){
      const self = this;
      this.componentConfig["datasets"].forEach(function(item){
        if (item["backgroundColor"] === undefined){
          item["backgroundColor"] = self.getThemeColor("primary")
        }else{
          if(Array.isArray(item["backgroundColor"])){
            item["backgroundColor"] = self.getThemeColorArray(item["backgroundColor"])
          }else{
            item["backgroundColor"] = self.getThemeColor(item["backgroundColor"])
          }
        }
        item["hoverBackgroundColor"] = "rgba(0, 0, 0, 0.1)"
      })
      this.chartJsInstance = new Chart(chartElement, {
        type: 'pie',
        data: {
            labels: this.componentConfig["labels"],
            datasets: this.componentConfig["datasets"]
        },
        options: {
            legend: {
              display: this.componentConfig["display_legend"],
            }
        }
      });
    },
  },

  mounted: function() {
    const style = getComputedStyle(document.body);

    this.defaultColor = style.getPropertyValue('--bs-primary');
    this.fontColor = style.getPropertyValue('--bs-secondary');
    this.fontFamily = style.getPropertyValue('--bs-font-sans-serif');

    Chart.defaults.global.defaultFontColor = this.fontColor
    Chart.defaults.global.defaultFontFamily = this.fontFamily
    // Chart.defaults.global.defaultColor = this.fontFamily

    const chartElement = this.$refs.chart

    if(this.componentConfig["type"] == "bar"){
      this.drawBarChart(chartElement);
    }
    if(this.componentConfig["type"] == "line"){
      this.drawLineChart(chartElement);
    }
    if(this.componentConfig["type"] == "doughnut"){
      this.drawDoughnutChart(chartElement);
    }
    if(this.componentConfig["type"] == "pie"){
      this.drawPieChart(chartElement);
    }

  },

  created: function() {

  },

  beforeDestroy: function() {

  },
});

File: app/concepts/matestack/ui/bootstrap/components/chart.rb
class Matestack::Ui::Bootstrap::Components::Chart < Matestack::Ui::VueJsComponent
  vue_js_component_name "matestack-ui-bootstrap-chart"

  requires :type
  requires :datasets
  optional :labels
  optional :height
  optional :width
  optional :display_legend
  optional :display_x_axes
  optional :display_y_axes
  optional :cutout_percentage
  optional class: { as: :bs_class }

  def setup
    @component_config[:type] = type
    @component_config[:datasets] = datasets
    @component_config[:labels] = labels
    @component_config[:display_legend] = !display_legend.nil? ? display_legend : false
    @component_config[:display_x_axes] = !display_x_axes.nil? ? display_x_axes : true
    @component_config[:display_y_axes] = !display_y_axes.nil? ? display_y_axes : true
    @component_config[:display_y_axes] = !display_y_axes.nil? ? display_y_axes : true
    @component_config[:cutout_percentage] = !cutout_percentage.nil? ? cutout_percentage : 70
  end

  def response
    div class: "chart-container #{bs_class}", attributes: { style: "width: 100%; height: 100%;" } do
      plain "<canvas ref='chart'></canvas>".html_safe
    end
  end

end

# chart class: "w-50", type: :bar, datasets: [
#   {
#     label: "€",
#     data: [total, last_7_days, last_30_days],
#     backgroundColor: :primary
#   },
# ], labels: ["total", "last 7 days", "last 30 days"]

# chart type: :doughnut, datasets: [
#   {
#     label: "€",
#     data: [total, last_7_days, last_30_days],
#     backgroundColor: [:orange, :secondary, :primary]
#   },
# ], labels: ["total", "last 7 days", "last 30 days"]

# chart type: :line, datasets: [
#   {
#     label: "€",
#     data: [total, last_7_days, last_30_days],
#     borderColor: :primary,
#     pointRadius: 0
#   },
#   {
#     label: "€",
#     data: [last_7_days, total, last_30_days],
#     borderColor: :danger,
#     fill: false
#   },
# ], labels: ["total", "last 7 days", "last 30 days"], display_x_axes: false, display_y_axes: false

# chart type: :pie, datasets: [
#   {
#     label: "€",
#     data: [total, last_7_days, last_30_days],
#     backgroundColor: [:orange, :secondary, :primary]
#   },
# ], labels: ["total", "last 7 days", "last 30 days"]
